Hierarchical Causal Models

adamrupe opened the pull request:
Implementation of algorithms in Hierarchical Causal Models paper by E. N. Weinstein and D. M. Blei.

cthoyt commented on pyproject.toml:
```diff
@@ -214,6 +215,9 @@ ignore = [
 "src/y0/algorithm/separation/sigma_separation.py" = ["RUF002", "RUF003"]
 "src/y0/algorithm/separation/__init__.py" = ["RUF002", "RUF003"]
+"tests/**/*.py" = ["S101"]
+"tests/test_hierarchical.py" = ["N803", "N802", "N806"]
```
I'm going to remove these ignores before making a review - these linting rules are to make sure that we have good variable names that follow a convention. Please carefully read through the feedback and try and pick better variable names

cthoyt reviewed (changes requested):
I did some housekeeping in #268 to make the testing work better for this branch.

First round of major review:

## Linting

Please remove the relaxations in linting. All of the N-series linting help keep our code understandable by reminding us to use good variable names (I removed them in aa2979e so CI will automatically reflect these expectations).

Using the following code as an example:

```python
@pytest.fixture
def instrument_collapsed_nxmixedgraph() -> NxMixedGraph:
    """Pytest fixture for collapsed Instrument HCM in Figure 2 (k)."""
    Qaz = Variable("Q_{a|z}")
    Y = Variable("Y")
    Qz = Variable("Q_z")
    return NxMixedGraph.from_edges(undirected=[(Qaz, Y)], directed=[(Qaz, Y), (Qz, Y)])
```

Rather than calling this variable `Qaz`, maybe come up with a name that describes what's the purpose of this variable in this example. I can't give a suggestion here what to change it to, because the docstring is also lacking. Maybe you can also give a functional explanation for all of the fixtures as to what is the purpose of the example graph

## Typing

Please address all mypy feedback

## Testing Framework

Please switch all tests in `test_hierarchical.py` from pytest to unittest.TestCase and match the rest of the testing suite. Most of the places in this PR that are using more advanced features like fixtures can actually just be functions, or even better, create a `y0.hcm_examples` module where you can assign a proper name to each example, and make them reusable

Sorry I waited until the end to give this feedback, but I didn't want to distract you in the middle of your work

## Notebook

The CI tests make sure that the notebook can be fully re-run, and right now it can't. Please take a look at that and make appropriate updates so it can run from scratch.

cthoyt commented on tests/test_hierarchical.py:
```diff
@@ -0,0 +1,822 @@
+@pytest.fixture
+def instrument_augmented_nxmixedgraph() -> NxMixedGraph:
+    """Pytest fixture for augmented Instrument HCM in Figure A2."""
+    Qaz = Variable("Q_{a|z}")
```
Make all of these variables global in the test_hierarchical.py module so you don't have to keep re-defining them

cthoyt reviewed (changes requested):
I did a major refactor to address the software issues from the last round. The next steps for @adamrupe and @djinnome are:

1. Read through the code and familiarize yourselves with the new interface
2. Comment on / address all TODO's I left in the code (there aren't many)
3. Tests
   - Either implement tests for conversion to HSCM or delete the conversion code
   - Test `augment_collapsed_model`
5. Check the notebook, which used to raise some exceptions, but I replaced those with the high-level `identify_outcomes` API. Please review to make sure that the estimands produced are all correct
6. Create a high-level, real world example that demonstrates using all of the code in a story-driven workflow (i.e., do not explain the math, only explain which functions you implemented solve the problem). Use https://github.com/y0-causal-inference/y0/blob/main/notebooks/Counterfactual%20Transportability.ipynb as a golden standard for how a great notebook with applications looks

Along the way, please make sure that you check the CI/CD system for automated, objective feedback on code quality. @adamrupe if you're not familiar with how to do this, I am happy to show you

adamrupe commented:
@cthoyt What's your recommendation for handling merge conflicts with jupyter notebooks? I need to do this before I can pull your updates. I'm also not familiar with the CI/CD system, so if you could talk me through it that would be great.

cthoyt commented:
@adamrupe before merging, copy your local notebook to your desktop. While merging, throw away everything from your repository's copy and overwrite it with remote. Then, you can think about manually inspecting your notebook on your desktop, and the new version from the remote repo side-by-side.

The best way to avoid this kind of thing is never to leave changes unpushed when you finish working, and to always pull before you start working again

---

The short explanation of how to use the CI/CD system is: you can always scroll to the bottom of this pull request (https://github.com/y0-causal-inference/y0/pull/236) and look at the feedback given by GitHub running our unit tests, linting, and code quality checks.

This is what it looks like to me right now:

<img width="946" alt="Screenshot 2025-02-05 at 23 51 20" src="https://github.com/user-attachments/assets/43d00266-5779-4513-99c7-a9818862d3e6" />


You can click on any of the rows with the red x's, and then it will bring you to the page that ran the tests for you. Right now, you will be able to see all of the output from running `pytest`. You have to scroll up a bit since unfortunately, pytest reports timings and warnings after test failures, but you can see https://github.com/y0-causal-inference/y0/actions/runs/13134504627/job/36646756591?pr=236#step:6:69 for the currently failing test.

Similarly, while you're still getting used to having code quality checks, you will probably see that the linting or type checking scripts also give errors, which you can view in the same way..

It's sort of the expectation in a team setting for coding that you make pushes often, and each time check out what kind of feedback CI gives. This will help you iteratively make your code better, with fully objective feedback that you don't have to wait on someone else to give you. Alternatively to CI/CD in GitHub, you can run `tox` which also creates a reproducible execution of all of the testing suite.

There's documentation in the README on how to use all of the nice development tools built into this repo at https://github.com/y0-causal-inference/y0?tab=readme-ov-file#%EF%B8%8F-for-developers

If you get caught up on any parts of this that aren't self-explanatory, I'm happy to plan a video chat tomorrow, or sometime next before 6PM germany time

adamrupe commented:
Awesome, thanks @cthoyt! That makes sense, and Jeremy and Richard have already shown me how to use `tox` a bit. I've pulled your changes and I'm going through them now. I'll add a `test_to_hscm`.

adamrupe commented:
@cthoyt I've refactored HSCMs and filled in the HCM.to_hscm() tests, so all tests are now passing. However, there is a depreciation warning I'm getting from another part of the codebase:
```
src/y0/examples/__init__.py:1173: FutureWarning: 
Downcasting behavior in `replace` is deprecated and will be removed in a future version. 
To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. 
To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
    asia_df = pd.read_csv(ASIA_PATH).replace({"yes": 1, "no": -1})
```
There is also a mypy error from `graph.py`
```
mypy: commands[0]> mypy --ignore-missing-imports --strict src/ tests/test_hierarchical.py
src/y0/graph.py:480: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
```
Since I didn't write this code, I didn't want to make edits to fix these, but they seem straightforward fixes. 

I'll add a test for `augment_collapsed_model`, and Jeremy and I have discussed a high-level notebook that I'll create as well. This should then complete your requested changes. I'll ping again when they are all complete.

cthoyt commented:
@adamrupe you should be unblocked on the CI/CD pipeline now. looking forward to seeing a nice case study notebook, then we can finish this PR!

adamrupe commented:
@cthoyt @djinnome I've changed the name of the previous notebook to HCM Manuscript Figures.ipynb and added a new case study notebook called Hierarchical Causal Models.ipynb.

cthoyt commented:
I'd like you to consider what makes https://github.com/y0-causal-inference/y0/blob/main/notebooks/Surrogate%20Outcomes.ipynb a joy to read and try and take some lessons from it to improve the HCM notebook.

1. Write your notebook keeping in mind that you are the last human being who ever has to understand the math behind the implementation you wrote.
2. Imagine that all users of `y0` want to solve a real problem, and they are reading the documentation to understand how they can model their problem using the data structures and algorithms in y0. They do not appreciate:
   - Abstract headings. Name each case study by the actual problem it's about, not the archetypical HCM as named by the paper. E.g., "Confounder" -> "After-school Tutoring and Test Scores"
   - Prose written like a mathematician. Rather than writing "Consider a school district that is interested in understanding how effective after-school tutoring is at raising test scores.", write "A school district is interested in understanding how effective after-school tutoring is at raising test scores." Use this simple and straightforward language to tell a story, not drag the user through a convoluted proof. Avoid words like "suppose", "consider"
   - Abstract examples. Give something concrete in any place you're tempted to use a variable to represent a high-level concept. Explain the concrete reason you need to make the modeling choice based on the case study, then after you may explain the theory that corresponds to that choice
   - Cryptic variable names. Any time you use a one letter variable name, you make it harder for readers to follow the example. Why are average test scores using the variable $y$? Call it `score`!
   - Avoid cryptic notation. Do the bars on top of the variable help understand what's going on? If we're not using individual test scores, then what does this add?
Further, why are we sub-scripting with i? The explanation for the subscripting comes way too late.
   - Mixture of typography. I think that it's better to re-produce Figure 1 inside the notebook rather than using a mixture of visual styles and fonts.
4. Use Python naming conventions for all variables. `Scores`, `Tutoring`, `UnitConfounder` should be `scores`, `tutoring`, and `unit_confounder`
3. I'm not sure you need to explain the process of collapsing and augmenting. Isn't `y0` able to abstract this away? Is there a logical reason the reader needs to know this happens in the background, given "here's the problem, here's how to model it, and here's the algorithm to apply to get an answer"? Maybe you can reuse some of the thoughts from above to frame this in a way that it's about the problem instead of about the math, but I think 
5. If you want to include math, you do have to address the difference between continuous integrals and the discrete estimands that come out of y0 functions. Further, try and match your hand-written notation to the output of y0 (e.g., use capital P for probability distributions, capital Q for Q-variables)

adamrupe commented:
@cthoyt thanks for the suggestions. Here are some thoughts, in no particular order: 

- We can add our own renderings instead of displaying Figure 1 from the paper. But there will still be a mix of typography because the hierarchical models require `pygraphviz`. It would be nice to have a pygraphviz backend for `NxMixedGraph.draw()`; in this case it would match styles, but `pygraphviz` generally has nice visuals. 
- On a related note, one reason for using abstract variable names like `Y` and `A` is that render nicely with `NxMixedGraph.draw()`. More verbose names like `scores` etc. do not fit in the nodes when visualized. 
- Speaking of Variable names, I used capitalized names for `y0` Variables, like `Scores` to follow the same convention used in other `y0` notebooks, including the Surrogate Outcomes notebook you have suggested (e.g. `Cancer`, `Smoking`, `Tar`). 
- Something very important to note is that the ideas and algorithms of hierarchical causal modeling are still in development. As of yet, there is no general and sound algorithm to answer whether a given causal query is identifiable from a hierarchical causal model. Therefore, in order to use the graphical algorithms we have implemented for hierarchical causal models, users need to have some baseline understanding of the algorithms and the ideas behind them, like what a Q Variable is. I've tried to distill out this baseline understanding from the (quite long) paper into the notebook. We can try to cut back on the math a little bit, but not too much. 
- For example, you asked whether the overbars are necessary to convey averaged quantities. In the first Confounder example, the distinction between quantities averaged over students in a school vs quantities for each individual student in the school is hugely important. The causal query is not identifiable when using the average values (which is not a hierarchical problem), but does become identifiable when we have the individual student data (and it is now a hierarchical problem). It is important for a user to understand why this is the case if they want to use these algorithms on their own hierarchical causal problems.

cthoyt commented:
fyi @adamrupe @djinnome I'll be starting my new job in RWTH aachen probably the third week of June, but as soon as I sign the contract (sometime between now and then), I plan to submit the y0 paper to JOSS. If we want to credit the implementation of HCMs in that paper, I would like this to be finished. Please let me know if it's not clear what the expectations are for finishing this / merging this PR

djinnome commented:
Congrats on the new position!  June is a good deadline for us. @adamrupe is not at PNNL anymore, but I have some time in May to finish off the remaining items

djinnome commented:
Referring back to https://github.com/y0-causal-inference/y0/pull/236#issuecomment-2744230646

I think we should help set reasonable expectations for the user.
The hierarchical causal model paper does not describe a sound and complete algorithm.
Certain decisions are still left to the user in terms of how to augment a hierarchical causal model so that it can be identified.
The confounder, Instrumental variables, and interference scenarios all demonstrate how to identify causal queries over hierarchical models that cannot be identified in a non-hierarchical model.
With that said, we are very much open to adjusting the notation and examples so that they are clearer to the user, even if the mechanics of the algorithms can't be completely ignored by the user.

cthoyt commented:
@adamrupe @njmmerrill there are only two tests failing, do you think we can address those and merge this (this PR is not blocked by finishing notebooks)?

cthoyt commented:
great! I split all of the non-notebook code out of this PR into a new PR, then merged it. I will leave it to all of you to continue work on the notebooks here.

If nobody intentionally changed the counterfactual transport notebook, then I would suggest deleting the `notebooks/Counterfactual Transportability edited.ipynb` file from this PR.
